File: day13/day13.cpp
#include <algorithm>
#include <fstream>
#include <climits>
#include <vector>
#include <fmt/format.h>

using namespace std;

static vector<int> parse_busses(const string& busses)
{
	vector<int> bids;
	size_t pos = 0, next;
	do {
		next = busses.find(',', pos);
		if (busses.substr(pos, next-pos) == "x")
		{
			bids.push_back(0);
		}
		else
		{
			bids.push_back(stoll(busses.substr(pos, next-pos)));
		}
		pos = next + 1;
	} while (next != string::npos);
	return bids;
}

static int part1(int earliest, const vector<int>& bids)
{
	int wait = INT_MAX;
	int bestid = 0;
	for (auto bid: bids)
	{
		if (bid)
		{
			int time = bid-1 - ((earliest + bid - 1) % bid);
			if (wait > time)
			{
				wait = time;
				bestid = bid;
			}
		}
	}
	return bestid * wait;
}

static intmax_t part2(const vector<int>& bids)
{
	vector<pair<int, int>> nums;
	for (int i = 0; i < (int)bids.size(); i++)
	{
		auto n = bids[i];
		if (n)
		{
			// NOTE: mathematically correct modulo of
			// negative numbers
			auto a = ((n - i) % n + n) % n;
			nums.emplace_back(make_pair(n, a));
		}
	}

	sort(nums.begin(), nums.end(), [](auto a, auto b) {
		return a.first > b.first;
	});
	intmax_t p = 1;
	intmax_t x = 0;
	for (auto [n, a]: nums)
	{
		while (x % n != a)
		{
			x += p;
		}
		p *= n;
	}
	return x;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	size_t earliest;
	string busses;
	input >> earliest;
	getline(input, busses);
	getline(input, busses);
	input.close();
	if (input.fail())
	{
		fmt::print(stderr, "Cannot parse the data\n");
		return 1;
	}

	auto bids = parse_busses(busses);

	fmt::print("Part1: {}\n", part1(earliest, bids));
	fmt::print("Part2: {}\n", part2(bids));
	return 0;
}

File: day13/day13.py
#!/usr/bin/env python3

import sys

def wait_time(estimate, buses):
    estimate = int(estimate)
    wait = estimate
    bestid = None
    for bus in buses.split(","):
        if bus == "x":
            pass
        else:
            bus = int(bus)
            test = bus - 1 - ((estimate + bus - 1) % bus)
            if wait > test:
                wait = test
                bestid = bus

    return wait * bestid

# NOTE: search by sieving (CRT)
def sieve(seq):
    nums = []
    for t, s in enumerate(seq.split(",")):
        if s != "x":
            n = int(s)
            nums.append((n, (n-t)%n))

    nums.sort(key = lambda x: -x[0])
    p, x = 1, 0
    for n, a in nums:
        while x % n != a:
            x += p
        p *= n

    return x

# assert sieve("7,13,x,x,59,x,31,19") == 1068781
# assert sieve("17,x,13,19") == 3417
# assert sieve("67,7,59,61") == 754018
# assert sieve("67,x,7,59,61") == 779210
# assert sieve("67,7,x,59,61") == 1261476
# assert sieve("1789,37,47,1889") == 1202161486

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    estimate, buses = txt.strip().split("\n")
    print("Part1:", wait_time(estimate, buses))
    print("Part2:", sieve(buses))
